### PR TITLE
ARROW-4267: [C++/Parquet] Handle duplicate and struct columns in RowGroup reads

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

**Comment by xhochy** [comment]
@wesm @kszucs @pitrou Can anyone have a short look at this? This is the last patch I want to put into the 0.12.1 bugfix release. Once this is merged, I would start the release process.

File: cpp/src/parquet/arrow/reader.cc [hunk 1]
```diff
@@ -494,24 +494,30 @@ Status FileReader::Impl::ReadRowGroup(int row_group_index,
...
-  int num_columns = static_cast<int>(indices.size());
-  std::vector<std::shared_ptr<Column>> columns(num_columns);
+  // We only need to read schema fields which have columns indicated
+  // in the indices vector
+  std::vector<int> field_indices;
+  if (!ColumnIndicesToFieldIndices(*reader_->metadata()->schema(), indices,
+                                   &field_indices)) {
+    return Status::Invalid("Invalid column index");
+  }
+  int num_fields = static_cast<int>(field_indices.size());
```
Comment (pitrou) [inline cpp/src/parquet/arrow/reader.cc]: IIUC, the fix is that `field_indices.size()` can be different from `indices.size()`?

**Comment by pitrou** [comment]
The PR title mentions struct columns. Is there a test for that?

File: cpp/src/parquet/arrow/reader.cc (hunk 1 above)
Comment (xhochy) [reply cpp/src/parquet/arrow/reader.cc]: Yes, this is the same as we do in https://github.com/apache/arrow/blob/6c4118b274cadf044b2d0581401a018f0a438205/cpp/src/parquet/arrow/reader.cc#L548-L556

That fix there was not applied to the single RowGroup reads.

**Comment by xhochy** [comment]
Added a small test for structs, turns out there was still a lot to do. Got the per-rowgroup code now in line with the full table code and tried to limit code duplication.

**Comment by wesm** [comment]
looking

**Review by wesm — APPROVED** [review]
+1. I'm going to try to spend a lot of April or May working on Parquet to help get nested data working and other cleaning in this module...

**Comment by wesm** [comment]
I'll merge as soon as the Appveyor build makes a bit more progress

**Pull request closed without merging**